pkg/controller/sparkapplication/controller.go

```diff
@@ -175,66 +175,30 @@ func (c *Controller) Stop() {
 // Callback function called when a new SparkApplication object gets created.
 func (c *Controller) onAdd(obj interface{}) {
 	app := obj.(*v1alpha1.SparkApplication)
-	if shouldSubmit(app) {
-		glog.Infof("SparkApplication %s was added, enqueueing it for submission", app.Name)
-		c.enqueue(app)
-		c.recorder.Eventf(
-			app,
-			apiv1.EventTypeNormal,
-			"SparkApplicationAdded",
-			"SparkApplication %s was added, enqueued it for submission",
-			app.Name)
-	}
+
+	glog.Infof("SparkApplication %s was added, enqueueing it for submission", app.Name)
+	c.enqueue(app)
 }
 
 func (c *Controller) onUpdate(oldObj, newObj interface{}) {
 	oldApp := oldObj.(*v1alpha1.SparkApplication)
 	newApp := newObj.(*v1alpha1.SparkApplication)
 
-	// The resource version has not changed, nothing to do.
-	if oldApp.ResourceVersion == newApp.ResourceVersion {
-		return
-	}
-
-	if reflect.DeepEqual(oldApp.Spec, newApp.Spec) {
-		// The spec has not changed but the application is subject to restart if the application state has changed.
-		// If the application state remains the same, it doesn't make sense to even check the restart eligibility.
-		if oldApp.Status.AppState.State != newApp.Status.AppState.State && shouldRestart(newApp) {
-			c.handleRestart(newApp)
-		}
-		return
-	}
-
-	if oldApp.Status.DriverInfo.PodName != "" {
-		// Clear the application ID if the driver pod of the old application is to be deleted. This is important as
-		// otherwise deleting the driver pod of the old application if it's still running will result in a driver state
-		// update by the sparkPodMonitor with the driver pod phase set to PodFailed. This may lead to a restart of the
-		// application if it'c subject to a restart because of the failure state update of the driver pod. Clearing the
-		// application ID causes the state update regarding the old driver pod to be ignored in
-		// processSingleDriverStateUpdate because of mismatched application IDs and as a consequence no restart to be
-		// triggered. This prevents the application to be submitted twice: one from the restart and one from below.
-		c.updateSparkApplicationStatusWithRetries(newApp, func(status *v1alpha1.SparkApplicationStatus) {
-			status.AppID = ""
-		})
-	}
-
-	// Delete the driver pod and UI service of the old application. Note that deleting the driver pod kills the
-	// application if it is still running. Skip submitting the new application if cleanup for the old application
-	// failed to avoid potentially running both the old and new applications at the same time.
-	if err := c.deleteDriverAndUIService(oldApp, true); err != nil {
-		glog.Errorf("failed to delete the old driver pod and UI service for SparkApplication %s: %v",
-			oldApp.Name, err)
+	// The spec has changed. This is currently not supported as we can potentially miss this update
+	// and end up in an inconsistent state.
+	if !reflect.DeepEqual(oldApp.Spec, newApp.Spec) {
```

Comment:
Actually it seems some users are really using this feature, e.g., triggering restart by patching a `SparkApplication`, so I think we should keep it.

Reply:
I would say let's revert the changes here as I heard a couple of use of this.

Reply:
Done, I ended up adding a new invalidating state to handle this in the new state machine. Have a look and let me know if there are any questions